### ISSUE TITLE
Digital signature of Windows executables

### ISSUE THREAD

**Issue by eliandoran** [TriliumNext/Notes#22]
### Discussed in https://github.com/orgs/TriliumNext/discussions/9

<div type='discussions-op-text'>

<sup>Originally posted by **deajan** February 15, 2024</sup>
We've already discussed this in one of the issues in zadam's repo.

Windows executables should be signed in order to avoid smartscreen screens.
One solution would be to buy a EV certificate (about $270/year) for the organization, and have one member keep the physical token.

Other solution is I can offer my company's code signing certificate services, but this would mean that I'd need to sign the windows executables (new code signing certificates require physical tokens). Of course, I vouch for my company's certificate ^^
</div>

---

See https://github.com/zadam/trilium/issues/4411 for details.

**Comment by deajan** [comment]
@eliandoran Just ping me whenever you need to sign something.
The EV cert I'm using is solely used to sign my own opensource software, so I'll be happy if my investment is used more broadly ;)

**Comment by eliandoran** [comment]
@deajan , the time has come.

If I remember correctly, the main roadblock was the fact that the certificate signing requires manual intervention.

Do you think there's any chance we could integrate it with the CI? Something like have the CI wait for you to sign it locally on your device? 🤔

**Comment by deajan** [comment]
@eliandoran Glad the time has come ;)

To be honest, unless the windows CI is hosted by the org, I have no clue how to attach the USB key to it.
Also, everytime it's used, it asks for a password, so I don't see any good way to automatize this.

The only thing I can do is signing when artifacts are created, manually, and provide them on a ftp/http server for you guys to upload. I can pledge to do this quickly (less than 24h generally) and for all official windows releases.

This can be done until the org has enough money to buy a EV certificate (with the same limitations btw).

**Comment by eliandoran** [comment]
@deajan , there might be some black magic involved but I think it's doable. Could you please have a look at: https://medium.com/@joshualipson/ev-code-certificates-automated-builds-for-windows-6100fb8e8be6

It would involve setting up the device you are using for signing right now as a GitHub Runner. If this is something that you'd like or can do, it would be great. Maybe your other projects could benefit from the same automation as well.

**Comment by deajan** [comment]
@eliandoran Yes, that's basically what I meant, the CI needs to be hosted by the org instead of github.
The findings on the article is interesting because it can be password-less.

I can setup a Windows VM, attach the EV USB stick to it via usb passthru, and add the github runner executable to it.
To keep things safe, the machine would download the actions script, which would then request the password from another file/registry.

Do you use standard github-action runners or custom build ones today ?

I can setup a github runner this weekend, but in order to do so, I'd need to get temporary write permissions on the org, in order to shift the CI runner from github to that one.

**Comment by eliandoran** [comment]
@deajan , sounds great.

I've given you temporary full access to the org. Feel free to add a org-level runner here: https://github.com/organizations/TriliumNext/settings/actions/runners

**Comment by deajan** [comment]
@eliandoran So today I've setup automatic EV signing for my projects. Your link needed a bit more research, but I managed to modify my signing code [here](https://github.com/netinvent/windows_tools/commit/c9df547888f3d85479aa3dc8b766c30a0eded389)



Since keeping the EV password secure is quite important, I've managed to write a bit of code which AES encrypts the EV cert data [here](https://github.com/netinvent/npbackup/blob/main/npbackup/windows/sign_windows.py)

I've tested this code against my personal projects that require signature.
Now that I'm done, I'll implement this on the github runner.
Again, sorry for the delay.

**Comment by deajan** [comment]
Runner is now registered at TriliumNext org.
Would you mind walking me through https://github.com/TriliumNext/Notes/blob/develop/.github/workflows/release.yml ?
I'm not sure why there where the zip files are generated, but I definitly need to unzip them in order to be able to sign the executables.

**Comment by eliandoran** [comment]
Hi, @deajan ,

This is great news!

The release CI script doesn't actually do much. It's a matrix which goes through all the platforms, including Windows 64.
Each build is actually an installation of dependencies, followed by running `npm run make-electron`. The resulting artifacts are either uploaded to GitHub Actions as artifacts, or to a release. The packaging (.zip, .exe) is handled automatically.

The build itself is handled using Electron Forge, which makes distributing Electron-based applications quite simple, including with the Squirrel installer for Windows and .dmg images for macOS.

I think the secret lies not within modifying the release script, but modifying the Electron Forge configuration. In our case the config file is in `/forge.config.cjs`.

It appears that there is an option to trigger an external tool to be able to sign the application, as mentioned in https://github.com/electron/forge/issues/3617. 

Let me know if you need more information and I'll happily provide them. Good luck!

**Comment by deajan** [comment]
@eliandoran I am truly sorry for the delay. I pledge to make this happen until the end of the year (lacking free time).

**Comment by deajan** [comment]
Currently working on it. Give me until this weekend. If not done, I pledge to donate enough for you guys to buy a code signing cert ;)

**Comment by eliandoran** [comment]
@deajan , no worries.

**Comment by deajan** [comment]
@eliandoran So this made much more headaches than I hoped for ^^ 

There are some rules that make signing executables a bit complicated. For instance, one cannot use a RDP connection to a Windows Server, since it will render the EV Token disabled.

Also, having an automatic signer can make the EV Token unusable if a wrong password is given three times, so I had to implement a sort of safeguard (if an error log file is present, nothing will be signed until the error log file is cleared).

Also, I did have fun understanding that there are Github Action runners at org and repo level (so I thought mine disappeared).

Anyway, I've got an isolated VM with a single binary that runs like
```
c:\git\ev_signer_trilium.exe --executable c:\path\to\my\file\to\sign.exe
```

This will output exit code 0 upon good signature, and 1 on any other case.
There is a new branch in notes repo, called ev-signature with a workflow that runs on one of my gh runners.
Would you mind modifying the forge script to call that tool ? I don't speak java/typescript myself.

**Comment by deajan** [comment]
@eliandoran So, new release is out. Did you get to adapt the workflow for digital signature via the custom github runner ?

**Comment by eliandoran** [comment]
@deajan , unfortunately not yet. I've been struggling a bit with Electron Forge since it seems to sign the executables in the ZIP archive and the one in the Windows installer differently.

Nevertheless, the runner you have provided seems to work just fine.

You can track the progress in https://github.com/TriliumNext/Notes/pull/1005, I hope to get it running in a week or two.
